### DISCUSSION TITLE
No healthy upstream issue in Ambient mode v1.23.0

### DISCUSSION THREAD

**Discussion by sachintiptur** [istio/istio#52878] [Q&A]
We are trying to integrate Istio into our clusters for the first time and using Istio Ambient mode. When we enabled Ambient mode for 
one of our workloads, it was throwing error in reaching kubernetes API server. The corresponding ztunnel logs show below messages. When i checked ztunnel-config output, I am seeing that kubernetes service has 0 endpoints where as other services have correct number of endpoint. Can you please let me know if we are missing something in our configurations ?

```
2024-08-27T13:35:00.524752Z    warn    access    connection failed    src.addr=10.23x.xx.22:yyy dst.addr=10.2x.0.1:443 direction="outbound" error="no healthy upstrea │
│ 2024-08-27T13:35:00.536627Z    warn    access    connection failed    src.addr=10.23x.xx.22:yyy dst.addr=10.2x.0.1:443 direction="outbound" error="no healthy upstrea │
│ 
```   

ztunnel-config service output shows 0 endpoints for kubernetes service and not listed in workloads also. This is observed for all the service endpoints whose workloads are using host Network.  Is it expected output or are we missing something??

```
$istioctl ztunnel-config service

default                  kubernetes                                     10.2x.0.1    None     0/0 

$istioctl ztunnel-config workloads | grep kube-apiserver
$

``` 

Unable to share bug-report at this point of time due to internal compliance restrictions.                                                                                                                                                            │

**Comment by howardjohn** [comment]
What version isthis? There were recent bug fixes in later 1.22.x and 1.23.0

**Reply by sachintiptur** [reply]
we are using v1.23. And the behavior is inconsistent. We have 3 control plane nodes, sometimes Istioctl shows 0 and sometimes 1 or 2. For all other workloads with hostNetwork set to true, shows 0 workloads.

**Reply by howardjohn** [reply]
Got it, thanks for the extra info. This is likely not a known issue then. It would be super helpful to have a bug-report in this case -- is it possible to capture one and send in a DM to me on slack (same username)?

If not, it would be good to get `kubectl get endpointslices kubernetes -oyaml; kubectl get service kubernetes -oyaml` at least

**Reply by sachintiptur** [reply]
```
kubectl get endpointslices kubernetes -oyaml
addressType: IPv4
apiVersion: discovery.k8s.io/v1
endpoints:
- addresses:
  - 10.10.91.11
  conditions:
    ready: true
- addresses:
  - 10.10.91.12
  conditions:
    ready: true
- addresses:
  - 10.10.91.13
  conditions:
    ready: true
kind: EndpointSlice
metadata:
  creationTimestamp: "2024-08-14T09:31:14Z"
  generation: 23
  labels:
    kubernetes.io/service-name: kubernetes
  name: kubernetes
  namespace: default
  resourceVersion: "11245793"
  uid: be3bec5e-ee19-4748-98e5-96693afce3c7
ports:
- name: https
  port: 6443
  protocol: TCP

kubectl get service kubernetes -oyaml
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: "2024-08-14T09:31:14Z"
  labels:
    component: apiserver
    provider: kubernetes
  name: kubernetes
  namespace: default
  resourceVersion: "227"
  uid: 794cc77e-e67c-4f59-a1e2-22984724b47b
spec:
  clusterIP: 10.2x.0.1
  clusterIPs:
  - 10.2x.0.1
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
  ports:
  - name: https
    port: 443
    protocol: TCP
    targetPort: 6443
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}

```

**Reply by howardjohn** [reply]
Hm, the exact config works for me on 1.23.0. The hostnetwork issue is fixed in https://github.com/istio/istio/pull/52191 which is not in 1.23.

Maybe a full `bug-report` would help -- sending over slack DM is okay if needed to keep it private

**Reply by howardjohn** [reply]
I confirmed https://github.com/istio/istio/pull/52191/ fixes this. Will see if we can partially backport it

**Reply by sachintiptur** [reply]
Thank you so much for looking into it. We will wait for the fix. :)

**Reply by sachintiptur** [reply]
Will it be backported to 1.23?

**Comment by maggie44** [comment]
I am seeing this too, in version `1.23.2`. Ztunnel logs:
>2024-09-24T11:47:36.302689Z warn access connection failed src.addr=192.168.152.146:59652 dst.addr=10.43.0.1:443 direction="outbound" error="no healthy upstream: 10.43.0.1:443"



`istioctl ztunnel-config service`
```
default              kubernetes                                         10.43.0.1     None     0/0
```


I am trying to access the Kubernetes API from inside a pod and it is failing. I have a deployment that starts a pod on 4 nodes. Sometimes 3 out of the 4 are working ok, but 1 isn't; another occasion 2 are not working and 1 is. 

The problem seems to move around. When I restart the deployment, different pods on different nodes can then access than before. 

When trying to access manually from inside the pods that have the issue (same command works in the pods without the issue):

```
TOKEN=$(cat /var/run/secrets/kubernetes.io/serviceaccount/token)
wget --header="Authorization: Bearer $TOKEN" --no-check-certificate https://10.43.0.1:443/api/v1/namespaces/...... -O -
Connecting to 10.43.0.1:443 (10.43.0.1:443)
wget: error getting response: Connection reset by peer
```

**Reply by maggie44** [reply]
If I do this from inside the ztunnel pod that is logging the issues about no healthy upstream I get a response, suggesting the ztunnel pod can see the API and query it:

```
curl https://10.43.0.1:443 --insecure 
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "Unauthorized",
  "reason": "Unauthorized",
  "code": 401
```
